File: starters/java/Map.java
// The Map class gets the Tron map from the game engine, and provides some
// methods that allow you to look at it.
//
// You should not change any of the code in this file. It's just here for
// your convenience, to do the boring parts for you.

import java.io.*;

class Map {
    // Stores the width and height of the Tron map.
    private static int width, height;

    // Stores the actual contents of the Tron map.
    private static boolean[][] walls;

    // Stores the locations of the two players.
    private static java.awt.Point myLocation, opponentLocation;

    // Returns the width of the Tron map.
    public static int Width() {
	return width;
    }

    // Returns the height of the Tron map.
    public static int Height() {
	return height;
    }

    // Returns whether or not the given space is a wall. The coordinates are
    // zero-based. Spaces outside the Tron map are deemed to be walls.
    public static boolean IsWall(int x, int y) {
	if (x < 0 || y < 0 || x >= width || y >= height) {
	    return true;
	} else {
	    return walls[x][y];
	}
    }

    // My X location.
    public static int MyX() {
	return (int)myLocation.getX();
    }

    // My Y location.
    public static int MyY() {
	return (int)myLocation.getY();
    }

    // The opponent's X location.
    public static int OpponentX() {
	return (int)opponentLocation.getX();
    }

    // The opponent's Y location.
    public static int OpponentY() {
	return (int)opponentLocation.getY();
    }

    // Transmit a move to the game engine. 'direction' can be any sort of
    // string that indicates a direction. For example, "north", "North",
    // "n", "N".
    public static void MakeMove(String direction) {
	if (direction.length() <= 0) {
	    System.err.println("FATAL ERROR: empty direction string. You " +
			       "must specify a valid direction in which " +
			       "to move.");
	    System.exit(1);
	}
	int firstChar = (int)direction.substring(0, 1).toUpperCase().charAt(0);
	switch (firstChar) {
	case 'N':
	    MakeMove(1);
	    break;
	case 'E':
	    MakeMove(2);
	    break;
	case 'S':
	    MakeMove(3);
	    break;
	case 'W':
	    MakeMove(4);
	    break;
	default:
	    System.err.println("FATAL ERROR: invalid move string. The string must " +
			       "begin with one of the characters 'N', 'E', 'S', or " +
			       "'W' (not case sensitive).");
	    System.exit(1);
	    break;
	}
    }

    // Reads the map from standard input (from the console).
    public static void Initialize() {
	String firstLine = "";
	try {
	    int c;
	    while ((c = System.in.read()) >= 0) {
		if (c == '\n') {
		    break;
		}
		firstLine += (char)c;
	    }
	} catch (Exception e) {
	    System.err.println("Could not read from stdin.");
	    System.out.flush();
	    System.exit(1);
	}
	firstLine = firstLine.trim();
	if (firstLine.equals("") || firstLine.equals("exit")) {
	    System.exit(1); // If we get EOF or "exit" instead of numbers
	                    // on the first line, just exit. Game is over.
	}
	String[] tokens = firstLine.split(" ");
	if (tokens.length != 2) {
	    System.err.println("FATAL ERROR: the first line of input should " +
			       "be two integers separated by a space. " +
			       "Instead, got: " + firstLine);
	    System.err.flush();
	    System.exit(1);
	}
	try {
	    width = Integer.parseInt(tokens[0]);
	    height = Integer.parseInt(tokens[1]);
	} catch (Exception e) {
	    System.err.println("FATAL ERROR: invalid map dimensions: " +
			       firstLine);
	    System.err.flush();
	    System.exit(1);
	}
	walls = new boolean[width][height];
	boolean foundMyLocation = false;
	boolean foundHisLocation = false;
	int numSpacesRead = 0;
	int x = 0, y = 0;
	while (y < height) {
	    int c = 0;
	    try {
		c = System.in.read();
	    } catch (Exception e) {
		System.err.println("FATAL ERROR: exception while reading " +
				   "from stdin.");
		System.exit(1);
	    }
	    if (c < 0) {
		break;
	    }
	    switch (c) {
	    case '\n':
		if (x != width) {
		    System.err.println("Invalid line length: " + x +
				       "(line " + y + ")");
		    System.exit(1);
		}
		++y;
		x = 0;
	    case '\r':
		continue;
	    case ' ':
		walls[x][y] = false;
		break;
	    case '#':
		walls[x][y] = true;
		break;
	    case '1':
		if (foundMyLocation) {
		    System.err.println("FATAL ERROR: found two locations " +
				       "for player " +
				       "1 in the map! First location is (" +
				       myLocation.getX() + "," +
				       myLocation.getY() +
				       "), second location is (" + x + "," +
				       y + ").");
		    System.exit(1);
		}
		walls[x][y] = true;
		myLocation = new java.awt.Point(x, y);
		foundMyLocation = true;
		break;
	    case '2':
		if (foundHisLocation) {
		    System.err.println("FATAL ERROR: found two locations for player " +
				       "2 in the map! First location is (" +
				       opponentLocation.getX() + "," +
				       opponentLocation.getY() + "), second location " +
				       "is (" + x + "," + y + ").");
		    System.exit(1);
		}
		walls[x][y] = true;
		opponentLocation = new java.awt.Point(x, y);
		foundHisLocation = true;
		break;
	    default:
		System.err.println("FATAL ERROR: invalid character received. " +
				   "ASCII value = " + c);
		System.exit(1);
	    }
	    ++x;
	    ++numSpacesRead;
	}
	if (numSpacesRead != width * height) {
	    System.err.println("FATAL ERROR: wrong number of spaces in the map. " +
			       "Should be " + (width * height) + ", but only found " +
			       numSpacesRead + " spaces before end of stream.");
	    System.exit(1);
	}
	if (!foundMyLocation) {
	    System.err.println("FATAL ERROR: did not find a location for player 1!");
	    System.exit(1);
	}
	if (!foundHisLocation) {
	    System.err.println("FATAL ERROR: did not find a location for player 2!");
	    System.exit(1);
	}
    }

    // Writes the given integer (direction code) to stdout.
    //   1 -- North
    //   2 -- East
    //   3 -- South
    //   4 -- West
    private static void MakeMove(int direction) {
	System.out.println(direction);
	System.out.flush();
    }
}

File: starters/clojure/map.clj
;; Map =================================================================
;;
;;  translated from the Java version by Nicolas Buduroi

(def width (ref 0))
(def height (ref 0))
(def walls (ref nil))

(defn point [x y] (vector x y))
(defn X [p] (p 0))
(defn Y [p] (p 1))

(defn init-map [w h]
  (dosync
   (ref-set width w)
   (ref-set height h)
   (ref-set walls {})))

(defn wall? [x y]
  (if (or (< x 0) (< y 0)
          (>= x @width) (>= y @height))
    true
    (@walls (point x y))))

(defn set-wall [x y]
  (dosync (alter walls assoc (point x y) true)))

(def players (ref (vec (repeat 2 #(point 0 0)))))

(defn you []  (first  @players))
(defn them [] (second @players))

(defn set-player [n x y]
  (dosync (alter players assoc n (point x y))))

(def directions
     {:north 1
      :east  2
      :south 3
      :west  4})

(defn exit [n] (System/exit n))

(defn error [& message]
  (.println *err* (apply str "FATAL ERROR: " message))
  (.flush *err*)
  (exit 1))

(defn check-game-over [input]
  (when (re-seq #"^(|exit)$" input)
    (exit 1)))

(defn check-first-line [input]
  (when (= 2 (count input))
    (error "The first line of input should be two integers separated "
           "by a space. Instead, got: " input)))

(defn parse-int [s]
  (try (Integer/parseInt s)
       (catch NumberFormatException e
         (error "Can't parse: " s))))

(defn check-line-length [x y]
  (when-not (= x @width)
    (error "Invalid line length: " x " at line " y " should be " @width)))

(defn check-player-found [players-found key]
  (when (players-found key)
    (error "Found two locations for player " key " in the map!")))

(defn check-player-missing [players-found key]
  (when-not (players-found key)
    (error "Did not find a location for player " key "!")))

(defn check-spaces-read [n]
  (when-not (= n (* @width @height))
    (error "Wrong number of spaces in the map.")))

(defn parse-first-line []
  (when-let [line (read-line)]
    (let [first-line (.trim line)
          tokens (seq (.split first-line " "))]
      (check-game-over first-line)
      (check-first-line first-line)
      (apply init-map (map parse-int tokens)))))

(defn initialize []
  (let [read #(.read *in*)]
    (try
     (parse-first-line)
     (loop [x 0 y 0
            players-found {}
            number-of-spaces 0]
       (if (>= y @height)
         (do (check-spaces-read number-of-spaces)
             (check-player-missing players-found \1)
             (check-player-missing players-found \2))
         (let [c (read)]
           (condp = (char c)
             \newline (do (check-line-length x y)
                          (recur 0 (inc y) players-found number-of-spaces))
             \return  (recur x y players-found number-of-spaces)
             \space   (recur (inc x) y players-found (inc number-of-spaces))
             \#       (do (set-wall x y)
                          (recur (inc x) y players-found (inc number-of-spaces)))
             \1       (do (check-player-found players-found \1)
                          (set-player 0 x y)
                          (recur (inc x) y (assoc players-found \1 true) (inc number-of-spaces)))
             \2       (do (check-player-found players-found \2)
                          (set-player 1 x y)
                          (recur (inc x) y (assoc players-found \2 true) (inc number-of-spaces)))
             (if (< c 0)
               (exit 0)
               (error "Invalid character received: " (char c)))))))
     (catch Exception e (error "Unknown exeption: " e)))))

(defn make-move [direction]
  (println (directions direction))
  (flush))
